README.md
# Timed URL Opener

Automatically open a specific URL at a scheduled time  in your browser 

![screencast](https://github.com/shufo/timed-url-opener/assets/1641039/6f2155e6-ab90-4921-a48c-297380c5e308)

## Features
- Open a specific URL at a scheduled time
  - Frequency by every xx minutes/hours/days, daily, weekday and One-time
- Open in Background option
- You can change Enbaled/Disabled per schedules
- You can change Color schemes (Dark Themes/Light Themes)

## Screenshot

<img height="300" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/a380ab13-39f5-4e26-a9a7-a328322eed57">
<img height="300" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/a311b23e-85cb-4e02-af2c-d57fead33d92">

## Installation

1. Go to [chrome webstore](https://chromewebstore.google.com/detail/timed-url-opener/calnlhiffgiakjlcfngkeljnplocbgpk) and install it.

## Usage

1. Press `Add Schedule` button.

    <img width="186" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/1ad56ead-c4a7-499a-985b-b76d92fb9623">

1. Enter URL you want to Open periodically.

    <img width="463" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/7b3a778f-508f-470f-9361-5d8dc983d516">

1. Specify time frequency and specific time.

    <img width="365" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/2e557347-5032-475a-9c89-ead2fff08a92">

1. If you want to delete schedule, press `Delete` button.

    <img width="393" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/591c0df2-e65d-4700-bffd-514bf3f30d80">

### Theme

You can change theme from sidebar settings

<img width="600" alt="image" src="https://github.com/shufo/timed-url-opener/assets/1641039/7f0f71ec-e7a9-4306-9451-ab2210523f45">

|Light|Nord|Valentine|Retro|Lofi|Cupcake|
|--|--|--|--|--|--|
|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/e8438d12-8561-4d53-b7d0-6973d1645b5e">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/ca933492-e14c-4983-813d-2e253af63f18">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/2c7961a9-7eb9-42bd-8e7e-a578adf52d38">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/2523d004-2077-44d2-bc28-a2cd3177752f">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/056f5c87-aded-4217-86ce-0d7e10f84cfd">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/1b3efe0b-074d-4430-a5a2-cd85c6d484a5">|

|Dark|Dracula|Dim|Synthwave|Sunset|Coffee|
|--|--|--|--|--|--|
|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/64e8cdda-fed7-4618-981b-b7a7a8883374">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/fac8fd35-bae8-425f-a673-ff4e6e09f601">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/5fd5db96-0c39-4825-b358-6454b5190c20">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/ba3a6a96-c322-45a9-858f-ba0c295a4369">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/5cc74561-2994-4c4c-9de0-bba7a85c0aef">|<img width="100" src="https://github.com/shufo/timed-url-opener/assets/1641039/7ed63eb7-d79b-46ff-b287-df2666a79966">|

## Development

```bash
$ npm install
# or 
$ yarn install
```

```bash
$ npm run dev
# or
$ yarn run dev
```

then Load `build/chrome-mv3-dev` directory on Extension manage page by enabling developer mode.

## production build

```bash
$ npm run build
# or
$ yarn run build
```

It will outputs production build to `build/chrome-mv3-prod` directory.

## Contributing

1.  Fork it
2.  Create your feature branch (`git checkout -b my-new-feature`)
3.  Commit your changes (`git commit -am 'Add some feature'`)
4.  Push to the branch (`git push origin my-new-feature`)
5.  Create new Pull Request

## LICENSE

MIT


CHANGELOG.md
# Changelog

## [1.0.2](https://github.com/shufo/timed-url-opener/compare/v1.0.1...v1.0.2) (2023-12-14)


### Bug Fixes

* 🐛 service worker unexpectedly stops after short period ([cd6f023](https://github.com/shufo/timed-url-opener/commit/cd6f023f425b21cb69a6aa0ec6f35385ebc2733e))

## [1.0.1](https://github.com/shufo/timed-url-opener/compare/v1.0.0...v1.0.1) (2023-12-12)


### Bug Fixes

* 🐛 delete unnecessary file ([7c3e65a](https://github.com/shufo/timed-url-opener/commit/7c3e65aa6a7a6cb0ab20b7cbdb595ce866c1ac36))

## 1.0.0 (2023-12-12) First release


### Features

* 🎸 display badge text and title if enbaled schedule exists ([eb5563a](https://github.com/shufo/timed-url-opener/commit/eb5563af4f80d7dcb8b458fed294e4a9344937df))


### Bug Fixes

* 🐛 delete unnecessary tabs permission request ([c3158f0](https://github.com/shufo/timed-url-opener/commit/c3158f012d13805050f9170136b7ab4bc66ef271))
* **deps:** update dependency tailwindcss to v3.3.6 ([4bbb30b](https://github.com/shufo/timed-url-opener/commit/4bbb30b4884d7b274d9f32c30ded856b986f2c4c))


tests/utils.ts
import dayjs from "dayjs";
import { ScheduleInput } from "~options";
import { generateUUID } from "~utils";

export const factoryInput = (options = {}) => {
    const input: ScheduleInput = {
        enabled: true,
        title: "",
        url: "",
        time: dayjs().add(5, "minutes").format("HH:mm"),
        date: dayjs().format("YYYY-MM-DD"),
        frequency: "daily",
        uuid: generateUUID(),
        amount: 1,
        unit: "minutes",
        background: false,
        added_at: new Date().toISOString(),
        is_valid: false,
        ...options
    }

    return input
}

src/utils.ts
import dayjs, { Dayjs } from "dayjs"
import type { ScheduleInput, Unit } from "~options"

export const generateUUID = () => {
    // Public Domain/MIT
    var d = new Date().getTime() //Timestamp
    var d2 = (performance && performance.now && performance.now() * 1000) || 0 //Time in microseconds since page-load or 0 if unsupported
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
        /[xy]/g,
        function (c) {
            var r = Math.random() * 16 //random number between 0 and 16
            if (d > 0) {
                //Use timestamp until depleted
                r = (d + r) % 16 | 0
                d = Math.floor(d / 16)
            } else {
                //Use microseconds since page-load if supported
                r = (d2 + r) % 16 | 0
                d2 = Math.floor(d2 / 16)
            }
            return (c === "x" ? r : (r & 0x3) | 0x8).toString(16)
        }
    )
}

export const getCloseTime = (unit: Unit, amount: number, startTime: Dayjs) => {
    const now = dayjs()
    const diffInMinutes = now.diff(startTime, "minute")

    let surplus = 0

    switch (unit) {
        case "minutes":
            surplus = diffInMinutes % amount
        case "hours":
            surplus = diffInMinutes % (amount * 60)
        case "days":
            surplus = diffInMinutes % (amount * 60 * 24)
    }

    if (surplus > 0) {
        return dayjs().second(0).subtract(surplus, "minute")
    }

    return startTime.second(0).add(amount, "minute")
}

export const generateNext5Dates = (input: ScheduleInput): Dayjs[] => {
    // daily
    if (input.frequency === "daily") {
        const dates = []

        const hour = parseInt(input.time.split(":")[0])
        const minute = parseInt(input.time.split(":")[1])
        const startTime = dayjs().hour(hour).minute(minute).second(0)

        for (let i = 0; i < 5; i++) {
            if (startTime.isAfter(dayjs())) {
                dates.push(startTime.add(i, "day"))
                continue
            }

            dates.push(startTime.add(i + 1, "day"))
        }

        return dates
    }

    // weekday
    if (input.frequency === "weekday") {
        const dates = []

        const hour = parseInt(input.time.split(":")[0])
        const minute = parseInt(input.time.split(":")[1])
        let startTime = dayjs().hour(hour).minute(minute).second(0)

        // get next 5 weekdays
        while (dates.length < 5) {
            if (startTime.day() !== 0 && startTime.day() !== 6) {
                if (startTime.isAfter(dayjs())) {
                    dates.push(startTime)
                }
            }

            startTime = startTime.add(1, "day")
        }

        return dates
    }

    if (input.frequency === "oneshot") {
        const dates = []

        const year = parseInt(input.date.split("-")[0])
        const month = parseInt(input.date.split("-")[1])
        const day = parseInt(input.date.split("-")[2])
        const hour = parseInt(input.time.split(":")[0])
        const minute = parseInt(input.time.split(":")[1])

        dates.push(
            dayjs()
                .year(year)
                .month(month - 1)
                .date(day)
                .hour(hour)
                .minute(minute)
                .second(0)
        )

        return dates
    }

    let now = dayjs()

    const year = parseInt(input.date.split("-")[0])
    const month = parseInt(input.date.split("-")[1])
    const day = parseInt(input.date.split("-")[2])
    const hour = parseInt(input.time.split(":")[0])
    const minute = parseInt(input.time.split(":")[1])

    now = now
        .year(year)
        .month(month - 1)
        .date(day)
        .hour(hour)
        .minute(minute)
        .second(0)

    const dates = []

    // generate dates while next 5 days from now is reached
    let i = 0

    const current = dayjs()

    let currentTime = getCloseTime(input.unit, input.amount, now)

    while (dates.length < 5) {
        if (currentTime.isBefore(current)) {
            currentTime = currentTime.add(input.amount, input.unit)
            continue
        }

        dates.push(currentTime)

        currentTime = currentTime.add(input.amount, input.unit)
    }

    return dates
}

export function computeNextAlarmTime(time: string) {
    const now = new Date();
    const [hours, minutes] = time.split(':');
    const scheduledTime = new Date(now.getFullYear(), now.getMonth(), now.getDate(), parseInt(hours), parseInt(minutes));

    if (scheduledTime <= now) {
        scheduledTime.setDate(now.getDate() + 1);
    }

    return scheduledTime.getTime();
}

export function computeNextFiveWeekdayAlarmTime(time: string): Dayjs[] {
    const now = dayjs();
    const [hours, minutes] = time.split(':');
    let scheduledTime = dayjs(now).hour(parseInt(hours)).minute(parseInt(minutes)).second(0);

    if (scheduledTime.isBefore(now)) {
        scheduledTime = scheduledTime.add(1, 'day');
    }

    const dates = []

    while (dates.length < 5) {
        if (scheduledTime.day() !== 0 && scheduledTime.day() !== 6) {
            dates.push(scheduledTime);
        }
        scheduledTime = scheduledTime.add(1, 'day');
    }

    return dates
}

export const capitalize = (str) => {
    if (typeof str !== "string" || !str) return str
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase()
}

src/config.ts
import { BugIcon, Icon, MarkGithubIcon, ReportIcon, StarIcon } from "@primer/octicons-react";

type MenuIcon = {
    name: string;
    title: string;
    icon: Icon;
    href: string;
}

// menu bar icons for navigation
export const menuBarIcons: Record<string, MenuIcon> = {
    webstore: {
        name: "Web Store",
        title: "Rate this extension on chrome webstore",
        icon: StarIcon,
        href: "https://chrome.google.com/webstore/category/extensions",
    },
    bug: {
        name: "Bug",
        title: "Report a bug",
        icon: BugIcon,
        href: "https://github.com/shufo/timed-url-opener/issues/new"
    },
    feedback: {
        name: "Feedback",
        title: "Feedback",
        icon: ReportIcon,
        href: "https://github.com/shufo/timed-url-opener/issues/new"
    },
    github: {
        name: "GitHub",
        title: "Star this project on GitHub",
        icon: MarkGithubIcon,
        href: "https://github.com/shufo/timed-url-opener"
    },
};


src/background.ts
import { Storage } from "@plasmohq/storage"
import dayjs from "dayjs";
import type { ScheduleInput } from "~options";
import { computeNextAlarmTime, computeNextFiveWeekdayAlarmTime, generateNext5Dates } from "~utils";

// @ts-ignore
import activeIcon from "../assets/icon_active_32x32.png"
// @ts-ignore
import inactiveIcon from "../assets/icon_inactive_32x32.png"

const env = process.env.NODE_ENV;
const envPrefix = env === "development" ? "DEV | " : "";
const defaultTitle = `${envPrefix}Timed URL Opener`

const main = async () => {

    const storage = new Storage();

    const keepAlive = () => setInterval(chrome.runtime.getPlatformInfo, 20e3);

    chrome.runtime.onInstalled.addListener(async () => {
        await updateAlarms();
    });

    chrome.runtime.onStartup.addListener(async () => {
        await updateAlarms();
        keepAlive();
    });

    keepAlive();

    chrome.action.onClicked.addListener(function (tab) {
        chrome.runtime.openOptionsPage(() => console.log("options page opened"));
    })

    console.info("background script loaded")

    storage.watch({
        "inputs": async (c) => {
            await updateAlarms();
        },
        make: (c) => {
            console.log(c.newValue)
        }
    })

    async function updateAlarms() {
        chrome.alarms.clearAll();

        const inputs: ScheduleInput[] = await storage.get("inputs");

        await updateIcon(inputs);

        inputs.forEach(async (input) => {

            // return if url is empty
            if (!input.url) return;

            // return if url is not valid
            try {
                new URL(input.url);
            } catch (err) {
                return;
            }

            // return if time is empty when frequency is (daily, weekday, oneshot)
            if (!input.time && ["daily", "weekday", "oneshot"].includes(input.frequency)) return;

            // process by frequency
            switch (input.frequency) {
                case "every":
                    processEveryFrequency(input);
                    return;
                case "daily":
                    processDailyFrequency(input);
                    return;
                case "weekday":
                    processWeekdayFrequency(input);
                    return;
                case "oneshot":
                    processOneshotFrequency(input);
                    return;
                default:
                    return;
            }
        });
    }

    async function updateIcon(inputs: ScheduleInput[]) {
        // set icon to active if enabled inputs exists
        const active = inputs.some(({ url, enabled }) => url.length > 0 && enabled);

        if (active) {
            setIconActive()
        } else {
            setIconInactive()
        }

        // set badge text
        const enabledInputs = inputs.filter(({ url, enabled }) => url.length > 0 && enabled);
        const badgeText = enabledInputs.length > 0 ? enabledInputs.length.toString() : "";
        chrome.action.setBadgeText({ text: badgeText });

        // set title
        chrome.action.setTitle({ title: enabledInputs.length > 0 ? `${defaultTitle} (${enabledInputs.length} active schedules)` : defaultTitle });
    }

    function setIconActive() {
        chrome.action.setIcon({ path: activeIcon });
    }

    function setIconInactive() {
        chrome.action.setIcon({ path: inactiveIcon });
    }

    function processEveryFrequency(input: ScheduleInput) {
        const startTime = dayjs(`${input.date} ${input.time}`);
        const currentTime = dayjs();

        let periodInMinutes = 0;

        // convert amount to number if it is string
        if (typeof input.amount === "string") {
            input.amount = parseInt(input.amount);
        }

        switch (input.unit) {
            case "minutes":
                periodInMinutes = input.amount;
                break;
            case "hours":
                periodInMinutes = input.amount * 60;
                break;
            case "days":
                periodInMinutes = input.amount * 24 * 60;
                break;
        }

        const next5Dates = generateNext5Dates(input);

        chrome.alarms.create(`openURL${input.uuid}`, {
            when: startTime.valueOf() < currentTime.valueOf() ?
                next5Dates.at(0).valueOf() // if startTime is before currentTime, set alarm to next first date
                : startTime.valueOf(), // else set alarm to startTime
            periodInMinutes: periodInMinutes,
        });
    }

    function processDailyFrequency(input: ScheduleInput) {
        chrome.alarms.create(`openURL${input.uuid}`, {
            when: computeNextAlarmTime(input.time),
            periodInMinutes: 24 * 60,
        });
    }

    function processWeekdayFrequency(input: ScheduleInput) {
        const dates = computeNextFiveWeekdayAlarmTime(input.time);

        dates.forEach(async (date) => {
            await chrome.alarms.create(`openURL${input.uuid}${date.valueOf()}`, {
                when: date.valueOf(),
                periodInMinutes: 24 * 60 * 7,
            });
        });
    }

    function processOneshotFrequency(input: ScheduleInput) {
        chrome.alarms.create(`openURL${input.uuid}`, {
            when: computeNextAlarmTime(input.time),
        });
    }

    // open url when alarm is triggered
    chrome.alarms.onAlarm.addListener(async alarm => {
        const uuid = alarm.name.substring('openURL'.length);
        const inputs: ScheduleInput[] = await storage.get("inputs");

        // get property from inputs matched uuid
        const { enabled, url, background } = inputs.find(({ uuid: inputUUID }) => inputUUID === uuid);

        // do nothing if not enabled
        if (!enabled) return;

        const active = !background;
        // open url
        if (url) {
            chrome.tabs.create({ url, active });
            setIconActive()
        }
    });

};

main();